### ISSUE TITLE
🚀 Breaking changes for Yarn 4

### ISSUE THREAD

**Issue by arcanis** [yarnpkg/berry#3591]
This issue is just there to keep track of which changes we might want to do once we'll get to the next major release. Doesn't mean it'll be anytime soon 🙂 Will be updated as time passes and we think about more things.

- https://github.com/yarnpkg/berry/pull/4196
- https://github.com/yarnpkg/berry/pull/4218
- https://github.com/yarnpkg/berry/pull/4262
- https://github.com/yarnpkg/berry/pull/4261 (Removes `yup`)
- https://github.com/yarnpkg/berry/pull/4252
- https://github.com/yarnpkg/berry/pull/4254
- https://github.com/yarnpkg/berry/pull/4302
- https://github.com/yarnpkg/berry/pull/4299
- https://github.com/yarnpkg/berry/pull/4305 (uses `npm:` everywhere)
- https://github.com/yarnpkg/berry/pull/4253
- https://github.com/yarnpkg/berry/pull/3004
- https://github.com/yarnpkg/berry/pull/3691
- https://github.com/yarnpkg/berry/pull/4320
- https://github.com/yarnpkg/berry/pull/4589
- https://github.com/yarnpkg/berry/pull/4834
- https://github.com/yarnpkg/berry/pull/4524
- https://github.com/yarnpkg/berry/pull/4982
- [ ] Make commands git-aware again
    - [x] Make `yarn init` initialize a git repository (already the case in 3.x)
    - [ ] Make `yarn version` create tags for the relevant workspaces
- [ ] `Installer['getCustomDataKey']`: "Move this method into `Linker` so that linkers can use it to save some state useful to findPackageLocator (cf PnpmLinker)."
- [ ] Make `ReportError` lazily generate report messages when provided a configuration object by the streams; this would allow us to remove the `configuration` parameter from many helper functions where we only use it for display purposes.
- [x] Update `yarn workspaces foreach` recursive configuration; it's not clear at the moment what should happen when `-R` and `--since` are used together. Perhaps we'd need a `--recursive-dependents` and `--recursive-dependencies` flag?
- https://github.com/yarnpkg/berry/pull/4595
- [ ] Cleanup https://github.com/yarnpkg/berry/pull/3612
  - https://github.com/yarnpkg/berry/pull/4641
  - [ ] Discuss the `fetchPackageFromCache` options bag.
- [ ] Transfer `arcanis.vscode-zipfs` to the Yarnpkg org
- [x] Make Yarn detect whether it's running inside a public repository ([GitHub Actions](https://github.community/t/can-an-action-detect-that-it-is-running-against-a-private-repository/18080)) and, if it is AND the repository uses zero-installs, exit and recommend adding either `--check-cache` or `--no-check-cache`.
- [x] Write a formal spec for the PnP data files, so that 3rd-party resolvers (in particular native resolvers) can implement their own (originally I pushed for it to be a JS API so that we had more freedom to extend/correct the file format as needed, but in 3.x we didn't need to change it so it might now be stable enough that we can relinquish some control).
    - [x] https://github.com/yarnpkg/berry/pull/4671
    - [ ] ~~Consider only writing this data file by default, unless an option to generate an old-style `.pnp.cjs` file is set.~~ (nope)
- https://github.com/yarnpkg/berry/pull/4503
- https://github.com/yarnpkg/berry/pull/4639
- https://github.com/yarnpkg/berry/pull/4644
- https://github.com/yarnpkg/berry/pull/4645
- https://github.com/yarnpkg/berry/pull/4648
- https://github.com/yarnpkg/berry/pull/4649
- https://github.com/yarnpkg/berry/pull/4697
- https://github.com/yarnpkg/berry/pull/4698
- https://github.com/yarnpkg/berry/pull/4773
- https://github.com/yarnpkg/berry/pull/4774
- https://github.com/yarnpkg/berry/pull/4865
- https://github.com/yarnpkg/berry/pull/4898
- https://github.com/yarnpkg/berry/pull/4899
- https://github.com/yarnpkg/berry/pull/5604
- https://github.com/yarnpkg/berry/pull/5608
- https://github.com/yarnpkg/berry/pull/5619

**Comment by brillout** [comment]
>  Drop Node 12 support (EOL will be 2022-04-30)

This is a no-go for OSS devs. E.g. I run `vite-plugin-ssr`'s test suite against Node 12 to make sure `vite-plugin-ssr` works with Node 12.

**Comment by arcanis** [comment]
> This is a no-go for OSS devs. E.g. I run `vite-plugin-ssr`'s test suite against Node 12 to make sure `vite-plugin-ssr` works with Node 12.

We don't intend to support Node versions that aren't maintained anymore by Node themselves. Not only to make our life easier, but also because some of our dependency do that too, and need to be kept updated (thus making such Node requirements a necessity).

**Comment by brillout** [comment]
Ok makes sense. Even if v4 releases a couple of months earlier than `2022-04-30` that would be ok. Sorry for the noise.

**Comment by evanw** [comment]
From [this comment](https://github.com/evanw/esbuild/issues/1983#issuecomment-1029805127) it sounds like Yarn 4 might also change the data format:

> I'm considering switching to a pure JSON format in the next Yarn major (and we'd publish this JSON as a formal spec that third-party resolvers can rely upon)

I went looking for the corresponding issue for that and ended up here. Should that be on this list too?

**Comment by arcanis** [comment]
I added an item to the list 👍

**Comment by edvardchen** [comment]
> We'll be exploring a new resolution algorithm that will prevent most of the attacks similar to the recent [color.js hijacking](https://news.ycombinator.com/item?id=29851065). 
from https://dev.to/arcanis/yarn-32-libc-yarn-explain-next-major--o22

Is there any way to follow, the implementation or design or discussion?  @arcanis

**Comment by pcjmfranken** [comment]
The only installation instructions provided in the docs are for corepack. Since corepack is not always an option (security concerns, having no npm, etc.) it would be nice if the manual instructions could be added back.

**Comment by milesj** [comment]
@pcjmfranken Corepack doesn't require npm. It's built into Node.js as of v14.19, which everyone should be using since v12 is EOL in a few days.

**Comment by pcjmfranken** [comment]
@milesj Corepack is actually built into[^1] the NPM that ships with the official Node distributions. Even still, you can opt out of Corepack being included in custom builds.

You're right in that the majority of users should have it available but it's certainly not a given.

[^1]: You can see the files being added to the `node_modules` directory here: https://github.com/nodejs/node/blob/master/tools/install.py#L82-L121

**Comment by magnattic** [comment]
The default value change for `enableGlobalCache` should be well documented when yarn 4 is released. This tripped me up when switching since we sync our local cache folder into our docker dev environment and I was confused why docker didn't work anymore.

**Comment by wojtekxtx** [comment]
Id be good to be able to edit `yarnrc.json` file. Currently its uneditable.
![Zrzut ekranu 2022-12-3 o 19 23 20](https://user-images.githubusercontent.com/109670049/205456041-4721c638-bf6d-4998-91b6-988870290e77.png)

Just sayin'.........

**Comment by winterqt** [comment]
Looks like #4247, which bumped the cache key, was missed in this list.

**Comment by merceyz** [comment]
> Looks like #4247, which bumped the cache key, was missed in this list.

Bumping the cache key isn't a breaking change.

**Comment by winterqt** [comment]
Is it just a coincidence that the last released cache key bump was done in a major version, and included a changelog entry?

**Comment by devinrhode2** [comment]
Hey @arcanis, could any of these changes be cherry picked into v3?

**Comment by merceyz** [comment]
No, breaking changes can't be backported.

**Comment by PowerKiKi** [comment]
Yarn 4.0 [first rc release](https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F4.0.0-rc.1) is now over 1 year old. And there's been 43 rc releases since then. And the [4.0.0 milestone](https://github.com/yarnpkg/berry/milestone/2) has no other opened issue.

Is there anything that is blocking a final release of Yarn 4.0.0 ? Is there a rough ETA of when this might happen ?

I am aware that we can use rc releases in the meantime, but it does look a bit unusual to have such a long period of time for rc. If it is stable, then it should probably be released, but if it isn't, then we probably should populate the milestone to reflect what is missing, shouldn't we ?

**Comment by arcanis** [comment]
> Is there anything that is blocking a final release of Yarn 4.0.0 ?

Time! It's always time. Most of the work is done, but there's always a couple of smallish improvements we want to make around some "details". Myself I have a couple of things around JS constraints, and @merceyz mentioned he'd like to close https://github.com/yarnpkg/berry/issues/4952 first to make the migration smoother. There's also the website revamp, but it'll likely be shipped separately.

> Is there a rough ETA of when this might happen ?

Probably soon, but no ETA.

**Comment by PowerKiKi** [comment]
> Myself I have a couple of things around JS constraints

Could you create issues assigned to 4.0.0 for those things ? then the community might be able to help, at least a bit ?

**Comment by landsman** [comment]
Where I can find waiting issues to be done? Milestone is the right place? Because there is only this issues. https://github.com/yarnpkg/berry/milestone/2

**Comment by lockieluke** [comment]
everyone's been talking about how good yarn 4 is but where on earth are the yarn 4 builds

**Comment by adrian-gierakowski** [comment]
`yarn set version 4.0.0-rc.48`

**Comment by danielpza** [comment]
also `yarn set version canary` https://yarnpkg.com/cli/set/version

**Comment by lockieluke** [comment]
where can i find the change logs

**Comment by coolbaluk** [comment]
Dependabot updates seem broken due to the cache key change, and each PR will include a mass replacement of zips with vendoring enabled.

I know it's technically unrelated, but thought it's useful to have it documented here.

**Comment by arcanis** [comment]
Will they? From the look of it they use Corepack, so if your package manager is pinned via either `packageManager` or `yarnPath`, you should only get updated PR after you migrate to v4, right?

**Comment by coolbaluk** [comment]
> Will they? From the look of it they use Corepack, so if your package manager is pinned via either `packageManager` or `yarnPath`, you should only get updated PR after you migrate to v4, right?

@arcanis It would have been nice, but my guess is they are still using v3 even with the above configured. This is the behaviour I'm seeing:

![Screenshot 2023-08-28 at 19 16 47](https://github.com/yarnpkg/berry/assets/5834683/6a7e4486-90fe-4cfa-824e-4b810fa220a2)

With:

`"packageManager": "yarn@4.0.0-rc.50"` and `yarnPath: .yarn/releases/yarn-4.0.0-rc.50.cjs`

I will write an issue on their side.

**Comment by trivikr** [comment]
Since yarn 4.0 stable is released in https://github.com/yarnpkg/berry/releases/tag/%40yarnpkg%2Fcli%2F4.0.0, this issue can be closed.

**Comment by SimenB** [comment]
getting an overview of what breaking changes are relevant to different use cases of yarn would be nice, and not just a list of the commits

**Comment by arcanis** [comment]
We will have a blog post tomorrow getting into the high level changes. I'll keep the issue open until then (we just wanted to tag the release beforehand to avoid having to rush a fix if something broke during the deploy process).

**Comment by herzaso** [comment]
This release made all of builds to fail.

Here is an excerpt from one of our docker files:
```
FROM node:16.19.0-bullseye-slim

RUN yarn config set scripts-prepend-node-path true
RUN yarn set version 3.3.1
```

This image, seem to include the latest version of yarn (4 from several hours ago) and since we rely on 3.3.1 we try to set it.

This is the error that we see:
```
Step 7/18 : RUN yarn set version 3.3.1
 ---> Running in c2e50735b75c
Usage Error: This tool requires a Node version compatible with >=18.12.0 (got 16.19.0). Upgrade Node, or set `YARN_IGNORE_NODE=1` in your environment.
Yarn Package Manager - 4.0.0
```

Changing the line to `RUN YARN_IGNORE_NODE=1 yarn set version 3.3.1` does fix the issue, just thought you should know about it

**Comment by wojtekmaj** [comment]
@herzaso I think the error messages are clear enough. This is an expected behavior.

And do yourself a favor and upgrade Node.js to a version that has not reached end of life yet.

**Comment by herzaso** [comment]
@wojtekmaj as I said, the error was clear and we managed to fix, but enterprises move slow. We are in the process of moving to Node 20

**Comment by trusktr** [comment]
@arcanis Amazing! A few questions though, that aren't covered in the article, perhaps useful for people who are new to Yarn (maybe you can update the post):

- How do you upgrade to 4.0? There are no steps listed. What are the migration steps?
- Many warnings are gone (according to the article, haven't tried yet), but I thought they were useful because other package managers did not show them, so I enjoyed having more insight into my install setup. How do we restore it?
- The post mentions official plugins. However, I cannot find a page on the website describing the purpose of plugins (what does it allow people to do?), and what the list of official plugins and their descriptions is. The only pages I can find via the search bar are the `yarn plugin` CLI docs (does not explain plugins' purpose) and a tutorial on writing a plugin (but without context on what plugins enable, knowing how to make the shell for a plugin is not useful without knowing what to put inside).

**Comment by arcanis** [comment]
@trusktr this is interesting feedback, but since this thread was specifically to talk about the breaking changes in Yarn 4, I don't think we should start discussing the content of the blog post itself (there are 23+ other people subscribed to this thread, and I don't want to spam them; additionally, an issue is likely the wrong format since you have multiple points that would lead to interlaced and very hard to follow discussion).
